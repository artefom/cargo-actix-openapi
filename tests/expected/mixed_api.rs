#![allow(unused_imports)]

//! API auto-generated by apigen

use std::{
    collections::HashMap,
    fmt::{Debug, Display},
};

use serde::{Deserialize, Serialize};

use actix_web::{
    http::StatusCode,
    middleware::{NormalizePath, TrailingSlash},
    web, App, HttpRequest, HttpResponse, HttpServer, ResponseError,
};

use actix_web_prom::PrometheusMetricsBuilder;

use async_trait::async_trait;

// Defaults
// -------------------------------

// Enums
// -------------------------------

// Struct
// -------------------------------

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
pub struct GreetUserPath {
    /// The name of the user to greet.
    pub user: String,
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
pub struct GoodbyeUserPath {
    pub user: String,
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
pub struct GoodbyeUserPathV2 {
    pub user: i64,
}

// Error with details
// -------------------------------

/// Create detailed errors with ease
#[macro_export]
macro_rules! detailed {
    ($err:expr,$msg:expr) => {
        $crate::server::api::Detailed {
            error: $err,
            details: $msg.to_string(),
        }
    };
}

/// Bails with detailed api error
#[macro_export]
macro_rules! apibail {
    ($err:expr,$msg:expr) => {
        return Err($crate::server::api::Detailed {
            error: $err,
            details: $msg.to_string(),
        })
    };
}

pub trait StatusCoded {
    fn status_code(&self) -> StatusCode;
}

#[derive(Debug)]
pub struct Detailed<E> {
    pub error: E,
    pub details: String,
}

impl<E: Display> Display for Detailed<E> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}. Reason: {}", self.error, self.details)
    }
}

impl<E: Display + Debug> std::error::Error for Detailed<E> {}

impl<E: Display + Debug> ResponseError for Detailed<E>
where
    E: StatusCoded,
{
    fn status_code(&self) -> StatusCode {
        self.error.status_code()
    }
}

/// Converts some result to Result with detailed api error
pub trait ApiErr<T, E> {
    /// Wrap the error value with additional context.
    fn apierr<C>(self, err: C) -> Result<T, Detailed<C>>
    where
        C: Display + Send + Sync + 'static;
}

impl<T, E> ApiErr<T, E> for Result<T, E>
where
    E: Debug + Send + Sync + 'static,
{
    fn apierr<C>(self, err: C) -> Result<T, Detailed<C>>
    where
        C: Display + Send + Sync + 'static,
    {
        // Not using map_err to save 2 useless frames off the captured backtrace
        // in ext_context.
        match self {
            Ok(ok) => Ok(ok),
            Err(original_error) => Err(Detailed {
                error: err,
                details: format!("{:?}", original_error),
            }),
        }
    }
}

// Error
// -------------------------------

// Api service
// -------------------------------

#[async_trait(?Send)]
pub trait ApiService<S>
where
    S: Send + Sync + 'static,
{
    /// Returns a greeting to the user!
    async fn greet_user(
        data: web::Data<S>,
        path: web::Path<GreetUserPath>,
    ) -> web::Json<String>;
    /// Returns a greeting to the user!
    async fn goodbye_user(
        data: web::Data<S>,
        path: web::Path<GoodbyeUserPath>,
    ) -> web::Json<String>;
    /// Returns a greeting to the user!
    async fn goodbye_user_v2(
        data: web::Data<S>,
        path: web::Path<GoodbyeUserPathV2>,
    ) -> web::Json<String>;
}

// Run service function (+ helper functions)
// -----------------------------------------
static DOCS_OPENAPI: &str = include_str!("static/openapi_v1.yaml");
static DOCS_HTML: &str = include_str!("static/docs.html");
static DOCS_OPENAPIV2: &str = include_str!("static/openapi_v2.yaml");
async fn openapi() -> String {
    DOCS_OPENAPI.to_string()
}
async fn openapi_v2() -> String {
    DOCS_OPENAPIV2.to_string()
}
async fn docs() -> HttpResponse {
    HttpResponse::build(StatusCode::OK)
        .content_type("text/html; charset=utf-8")
        .body(DOCS_HTML)
}
async fn to_v1_docs() -> HttpResponse {
    HttpResponse::build(StatusCode::TEMPORARY_REDIRECT)
        .append_header(("Location", "v1/docs"))
        .body("")
}
async fn to_docs() -> HttpResponse {
    HttpResponse::build(StatusCode::TEMPORARY_REDIRECT)
        .append_header(("Location", "docs"))
        .body("")
}
async fn to_v2_docs() -> HttpResponse {
    HttpResponse::build(StatusCode::TEMPORARY_REDIRECT)
        .append_header(("Location", "v2/docs"))
        .body("")
}

// Tells that service is alive
async fn health() -> HttpResponse {
    HttpResponse::Ok().finish()
}

pub async fn run_service<T, S>(bind: &str, initial_state: S) -> Result<(), std::io::Error>
where
    T: ApiService<S> + 'static,
    S: Send + Sync + 'static,
{
    let app_data = web::Data::new(initial_state);

    let prometheus = PrometheusMetricsBuilder::new("api")
        .endpoint("/metrics")
        .build()
        .unwrap();

    use web::{delete, get, post};

    HttpServer::new(move || {

        let api = web::scope("")
            .wrap(prometheus.clone())
            .route("/goodbye/{user}", get().to(T::goodbye_user))
            .route("/hello/{user}", get().to(T::greet_user))
            .route("/v1/goodbye/{user}", get().to(T::goodbye_user))
            .route("/v1/hello/{user}", get().to(T::greet_user))
            .route("/v2/goodbye/{user}", get().to(T::goodbye_user_v2))
            .route("/v2/hello/{user}", get().to(T::greet_user))
            .wrap(prometheus.clone());

        App::new()
            .app_data(app_data.clone())
            .wrap(NormalizePath::new(TrailingSlash::MergeOnly))
            // Aux services
            .route("/health", get().to(health))
            // Static paths
            .route("/", get().to(to_v2_docs))
            .route("/docs", get().to(docs))
            .route("/openapi.yaml", get().to(openapi))
            .route("/v1", get().to(to_v1_docs))
            .route("/v1/", get().to(to_docs))
            .route("/v1/docs", get().to(docs))
            .route("/v1/openapi.yaml", get().to(openapi))
            .route("/v2", get().to(to_v2_docs))
            .route("/v2/", get().to(to_docs))
            .route("/v2/docs", get().to(docs))
            .route("/v2/openapi.yaml", get().to(openapi_v2))
            // Server routes
            .service(api)
    })
    .bind(bind)?
    .run()
    .await?;

    Ok(())
}
