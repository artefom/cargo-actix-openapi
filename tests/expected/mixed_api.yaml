api:
  definitions:
    DOCS_OPENAPI:
      data: !StaticStr
        path: static/openapi_v1.yaml
    DOCS_HTML:
      data: !StaticStr
        path: static/docs.html
    openapi:
      data: !StaticStringPath
        data: DOCS_OPENAPI
    docs:
      data: !StaticHtmlPath
        data: DOCS_HTML
    to_docs:
      data: !Redirect
        target: v1/docs
    GreetUserPath:
      data: !Struct
        doc: null
        properties:
        - name: user
          rename: user
          default: null
          type_: String
          doc: The name of the user to greet.
    GoodbyeUserPath:
      data: !Struct
        doc: null
        properties:
        - name: user
          rename: user
          default: null
          type_: String
          doc: null
    to_docs_v2:
      data: !Redirect
        target: v2/docs
    DOCS_OPENAPIV2:
      data: !StaticStr
        path: static/openapi_v2.yaml
    openapi_v2:
      data: !StaticStringPath
        data: DOCS_OPENAPIV2
    GoodbyeUserPathV2:
      data: !Struct
        doc: null
        properties:
        - name: user
          rename: user
          default: null
          type_: i64
          doc: null
  operations:
    greet_user:
      doc: Returns a greeting to the user!
      param_path: web::Path<GreetUserPath>
      param_query: null
      param_body: null
      response: web::Json<String>
    goodbye_user:
      doc: Returns a greeting to the user!
      param_path: web::Path<GoodbyeUserPath>
      param_query: null
      param_body: null
      response: web::Json<String>
    goodbye_user_v2:
      doc: Returns a greeting to the user!
      param_path: web::Path<GoodbyeUserPathV2>
      param_query: null
      param_body: null
      response: web::Json<String>
  paths:
  - operation: goodbye_user
    path: /goodbye/{user}
    method: Get
  - operation: greet_user
    path: /hello/{user}
    method: Get
  - operation: goodbye_user
    path: /v1/goodbye/{user}
    method: Get
  - operation: greet_user
    path: /v1/hello/{user}
    method: Get
  - operation: goodbye_user_v2
    path: /v2/goodbye/{user}
    method: Get
  - operation: greet_user
    path: /v2/hello/{user}
    method: Get
  static_services:
  - method: Get
    path: /
    data: to_docs_v2
  - method: Get
    path: /docs
    data: docs
  - method: Get
    path: /openapi.yaml
    data: openapi
  - method: Get
    path: /v1
    data: to_docs
  - method: Get
    path: /v1/docs
    data: docs
  - method: Get
    path: /v1/openapi.yaml
    data: openapi
  - method: Get
    path: /v2
    data: to_docs_v2
  - method: Get
    path: /v2/docs
    data: docs
  - method: Get
    path: /v2/openapi.yaml
    data: openapi_v2
